test/unittest/tests/test_assertions.cpp

```diff
@@ -13,39 +13,335 @@
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_OUT_OF_ORDER)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_SLOW_OPERATION)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_VISIBILITY)) == true);
+
+    return 0;
+}
+
+/* Assert that all asserts fire. */
+int
+all_asserts_abort(WT_SESSION_IMPL *session, uint16_t category, bool assert_should_pass)
+{
+    int ret = 0;
+
+    REQUIRE(assert_always_aborts(session) == 0);
+
+    REQUIRE(call_wt_optional(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_OPTIONAL raised assert");
+
+    REQUIRE(call_wt_ret(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_RET raised assert");
+
+    REQUIRE(call_wt_err(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_ERR raised assert");
+
+    REQUIRE(call_wt_panic(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_PANIC raised assert");
+
+    return ret;
+}
+
+/* Assert that the expected asserts fire. */
+int
+configured_asserts_abort(WT_SESSION_IMPL *session, uint16_t category, bool assert_should_pass)
+{
+    int ret = 0;
+
+    REQUIRE(call_wt_optional(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_OPTIONAL raised assert");
+
+    REQUIRE(call_wt_ret(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_RET raised assert");
+
+    REQUIRE(call_wt_err(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_ERR raised assert");
+
+    REQUIRE(call_wt_panic(session, category, 1 == 2) == 14);
+    expect_assertion(session, "Assertion '1 == 2' failed: WT_PANIC raised assert");
+
+    return ret;
+}
+
+/* Assert that the expected asserts don't fire. */
+int
+configured_asserts_off(WT_SESSION_IMPL *session, u_int16_t category, bool assert_should_pass)
+{
+    int ret = 0;
+
+    REQUIRE(call_wt_optional(session, category, 1 == 2) == 14);
+    expect_no_assertion(session);
+
+    REQUIRE(call_wt_ret(session, category, 1 == 2) == -1);
+    expect_no_assertion(session);
+
+    REQUIRE(call_wt_err(session, category, 1 == 2) == -1);
+    expect_no_assertion(session);
+
+    REQUIRE(call_wt_panic(session, category, 1 == 2) == -31804);
+    expect_no_assertion(session);
+
+    return ret;
+}
+
+/* Assert that regardless of connection configuration, asserts always disabled/enabled */
+TEST_CASE("Connection config: off", "[assertions]")
+{
+    ConnectionWrapper conn(DB_HOME, "create");
+    WT_SESSION_IMPL *session = conn.createSession();
+
+    REQUIRE(all_diag_asserts_off(session) == 0);
+}
+
+/* Assert that regardless of connection configuration, asserts always disabled/enabled */
+TEST_CASE("Connection config: on", "[assertions]")
```

Comment:
We can group this test case with `Connection config: off` above. 
It's also worth calling `assert_always_aborts` in that first test case

Reply:
These test cases were initially grouped but that would require a connection->reconfigure to switch the connection config from off->on, but these test cases are meant to test configuring when opening the connection and not using reconfigure, which is why I split the two test cases up. As discussed, I'll make this a single test_case with two different sections instead of grouping them together.